src/main/java/com/digitalasset/testing/utils/PackageUtils.java

```diff
@@ -0,0 +1,197 @@
+      return new TemplateType(identifiers.get(moduleAndEntityName), dt.getCreateFields(), m);
+    }
+    throw new IllegalArgumentException("No template found with the name " + moduleAndEntityName);
+  }
+
+  public static String dottedNameToString(DamlLf1.DottedName name) {
+    StringBuilder b = new StringBuilder();
+    for (int i = 0; i < name.getSegmentsCount(); i++) {
+      b.append(name.getSegments(i));
+    }
+    return b.toString();
+  }
+
+  private static DataType findDataType(DamlLedgerClient ledgerClient, String moduleAndEntityName)
+      throws InvalidProtocolBufferException {
+    assert !moduleAndEntityName.isEmpty();
+    DataType dt = dataTypes.get(moduleAndEntityName);
+    if (dt != null) {
+      return dt;
+    } else {
+      // Init or reinit the cache...
+      initCache(ledgerClient);
+      // Try again and throw.
+      dt = dataTypes.get(moduleAndEntityName);
+      if (dt != null) {
+        return dt;
+      }
+      throw new IllegalArgumentException(
+          "No datatype found with the name '" + moduleAndEntityName + "'");
+    }
+  }
+
+  private static void initCache(DamlLedgerClient ledgerClient)
+      throws InvalidProtocolBufferException {
+    PackageClient pkgClient = ledgerClient.getPackageClient();
+    Iterable<String> pkgs = pkgClient.listPackages().blockingIterable();
+    for (String pkgId : pkgs) {
+      GetPackageResponse pkgResp = pkgClient.getPackage(pkgId).blockingGet();
+      DamlLf.ArchivePayload archivePl =
+          DamlLf.ArchivePayload.parseFrom(pkgResp.getArchivePayload());
+      List<DamlLf1.Module> mods = archivePl.getDamlLf1().getModulesList();
+      for (DamlLf1.Module mod : mods) {
+        for (DamlLf1.DefDataType dataType : mod.getDataTypesList()) {
+          String modN = dottedNameToString(mod.getName());
+          String dataN = dottedNameToString(dataType.getName());
+          String moduleAndEntityName = toFqn(modN, dataN);
+          Identifier id = new Identifier(pkgId, modN, dataN);
+          identifiers.put(moduleAndEntityName, id);
+          dataTypes.put(moduleAndEntityName, new DataType(mod, dataType));
+        }
+      }
+    }
+  }
+
+  private static String toFqn(String moduleName, String entityName) {
+    Objects.requireNonNull(moduleName);
+    Objects.requireNonNull(entityName);
+    assert !moduleName.isEmpty() && !entityName.isEmpty();
+    return moduleName + ":" + entityName;
+  }
+
+  private static Map<String, DamlLf1.Type> getChoices(
+      DamlLf1.Module mod, DamlLf1.DottedName dataTypeName) {
+    Optional<DamlLf1.DefTemplate> template =
+        mod.getTemplatesList().stream().filter(t -> t.getTycon().equals(dataTypeName)).findFirst();
+    if (template.isPresent()) {
+      HashMap<String, DamlLf1.Type> m = new HashMap<>();
+      for (DamlLf1.TemplateChoice choice : template.get().getChoicesList()) {
+        m.put(choice.getName(), choice.getArgBinder().getType());
+      }
+      return m;
+    } else {
+      return null;
```

Comment:
just curiosity: why don't you return an empty map?

Reply:
I want to differentiate between no choices (empty map) and "this is not a template so having choices in unintelligible" (null).

Reply:
`Optional<Map<String, DamlLf1.Type>>` would be better then (all the way through the call-chain as much as needed)

Reply:
:+1: